### ISSUE TITLE
Exit 80131506 after installing Datadog csharp

### ISSUE THREAD

**Issue by nbosscher** [DataDog/dd-trace-dotnet#207]
**Describe the bug**
After installing Datadog c-sharp some of my applications won't start up, giving exit code `80131506`

**To Reproduce**
1. Install DataDog tracer from https://github.com/DataDog/dd-trace-csharp/releases
2. Restart IIS using steps on https://docs.datadoghq.com/tracing/setup/dotnet/
3. Load application
4. App Pool crashes before first page is loaded

**Expected behavior**
The application works the same after installing Datadog tracer.

**Screenshots**
None

**Runtime environment (please complete the following information):**
- Instrumentation mode: automatic with msi installer
- Tracer version: 0.5.0
- OS: Windows Server 2016 Data Center
- CLR: .NET 4.0.30319 (reported in IIS), .NET 4.7 installed on machine

**Additional context**
Event Viewer Logs
```
Application Error:
Faulting application name: w3wp.exe, version: 10.0.14393.0, time stamp: 0x57899b8a
Faulting module name: clr.dll, version: 4.7.3190.0, time stamp: 0x5b693b48
Exception code: 0xc0000005
Fault offset: 0x000000000056604b
Faulting process id: 0x4d8
Faulting application start time: 0x01d47a9f85863416
Faulting application path: c:\windows\system32\inetsrv\w3wp.exe
Faulting module path: C:\Windows\Microsoft.NET\Framework64\v4.0.30319\clr.dll
Report Id: 769c85ef-ad3c-45f2-9441-e4d4ce2a7381
Faulting package full name: 
Faulting package-relative application ID: 

.NET Runtime:
Application: w3wp.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an internal error in the .NET Runtime at IP 00007FFD0A77604B (00007FFD0A210000) with exit code 80131506.
```

Attempted to change GC settings, but it didn't help
- tried `<gcServer enabled="true" />`
- tried `<gcConcurrent enabled="false" />`

Oddly this doesn't happen for all of my .NET applications, some work fine.
- both applications are compiled with `.NET 4.5.2`

Debugging with sysinternals/procmon showed attempts to write to `ProgramData\Datadog .NET Tracer\logs`, but `ProgramData\Datadog .NET Tracer` didn't exist...?
- After manually creating `ProgramData\Datadog .NET Tracer\logs`, a `dotnet-profiler.log` was created, but only had `[info]` level messages in it

Going to try attaching with Windbg, but I'm a bit out of my depth on this.

**Comment by nbosscher** [comment]
Narrowed it down a bit. Here's a stack trace
```
clr.dll!ProfToEEInterfaceImpl::GetModuleFlags(class Module *)
clr.dll!ProfToEEInterfaceImpl::GetModuleInfo2(unsigned __int64,unsigned char const * *,unsigned long,unsigned long *,unsigned short * const,unsigned __int64 *,unsigned long *)
Datadog.Trace.ClrProfiler.Native.dll!00007ffcd9ec1bee()
```

**Comment by nbosscher** [comment]
By running the debugger and sysinternals/proc-mon, I was able to find that the last error before the crash was related to loading ClearScript. As I said, if I uninstall the DataDog .NET integration, the application loads ok, so I can only conclude that DataDog is not handling some error case correctly. Perhaps, you're requesting a module that didn't actually load properly...?

**Comment by lucaspimentel** [comment]
Hey, @nbosscher. Thanks for reporting this issue. The profiler writes it's log file to `ProgramData\Datadog .NET Tracer\logs` like you said. If that folder doesn't exist or there are no files there, then it's possible it didn't have enough permissions to create the folders and write to the log file.

The profiler is a native library (dll) that gets loaded into you application's process, so it runs with the same privileges as you application. Do you know if your application is somehow limited to writing to that folder?

You also mentioned ClearScript. The .NET Tracer does not use ClearScript. Is this a library your application uses?

Thanks again.

**Comment by nbosscher** [comment]
Thanks for the reply @lucaspimentel 

Yep, that's exactly right. It didn't have permissions for that directory. Perhaps that's something that got missed in the installer...

Yes, we use ClearScript in our application.

**Comment by lucaspimentel** [comment]
Thanks for confirming that it was a permissions issue. At the very least, we should catch that exception and not crash the profiled application because we were not able to write to the log file.

**Comment by nbosscher** [comment]
@lucaspimentel, unfortunately the log file is not the only issue I reported above. There's also the clearscript errors (.net clr exceptions)

**Comment by lucaspimentel** [comment]
@nbosscher, can you open a separate issue for the ClearScript error? Thanks.

**Comment by nbosscher** [comment]
Ok, opened #216

**Comment by lucaspimentel** [comment]
Oops. I asked you to create a separate issue because I thought most of those error details above were related to the file permissions issue, but it looks like it was all about ClearScript error. I'll close issue #216 since it's basically a duplicate of this one. Sorry about that.

**Comment by lucaspimentel** [comment]
I looked at ClearScript's NuGet package and noticed that it includes both 64-bit and 32-bit versions of V8's native libraries. A `targets` file copies all those DLLs to the output folder. I suspect the issue here is that somehow the runtime is triggered into trying to load those DLLs and it's choking either on the fact that they are not managed assemblies, or that it's trying to load a 32-bit DLL into a 64-bit process (or vice-versa). I'll do more testing...

**Comment by lucaspimentel** [comment]
@nbosscher: What you tell me the NuGet ID and version of the ClearScript package you are using? Also, is your app running in 32-bit or 64-bit mode?

**Comment by nbosscher** [comment]
@lucaspimentel Glad to see you understand my ticket a bit better now. Sorry if I structured it in a way that was confusing.

- ClearScript version: (from packages.config)
```
<package id="ClearScript.V8" version="5.4.6" targetFramework="net451" />
```
- The application is running in 64-bit mode

Idk if it's worth mentioning but we received a lot of ClearScript errors via ProcMon prior to the failure. 

I don't have time to work out a sample app for you right now. Basically our usage was that we're using ClearScript for the BundleTransformer package. On the application's first request it would crash the app pool. From the that stack trace I gave you, it basically looks like an invalid object-id is being passed to the .net clr and the clr is choking. I hope that you're able to figure this out with that information. 

If you run stuck, let me know and I'll see if I can get some time later this week to work a sample app out for you.

**Comment by lokalan** [comment]
Any progress being made on this issue?

I'm currently evaluating the .NET APM product and ran into this same issue, except oddly enough it works fine on IIS but it's a console application that crashes.  Both use the ClearScript package.

If I delete the 32 bit assemblies from the console app's bin folder (`v8-ia32.dll` and `ClearScriptV8-32.dll`, the app runs.

**Comment by lucaspimentel** [comment]
@lokalan: Aside from the obvious, are there any other differences between the IIS app and the console app? For example, is the console app is running in 64-bit mode while the IIS app is running in 32-bit mode? Have you tried running the console app in 32-bit mode? A minimal project that reproduces this issue would be ideal.

**Comment by lokalan** [comment]
They are both running in 64-bit mode.

After a bit of experimenting, I think I've narrowed it down to a 3rd party IoC library that's scanning all `dll` files in the application directory.

I created a console app that mimics what the 3rd party library does, note that even though we look for and catch `BadImageFormatExceptions`, when running under the Datadog APM profiler, it just hard crashes the process.  Running without the profiler does not throw any errors.

https://github.com/lokalan/DatadogApmClearscriptCrash
